src/main/user-impl/java/com/mysql/cj/jdbc/ha/MultiHostConnectionProxy.java

```diff
@@ -354,14 +360,16 @@ protected synchronized void invalidateConnection(JdbcConnection conn) throws SQL
      * @throws SQLException
      *             if an error occurs
      */
-    protected synchronized ConnectionImpl createConnectionForHost(HostInfo hostInfo) throws SQLException {
-        ConnectionImpl conn = (ConnectionImpl) ConnectionImpl.getInstance(hostInfo);
-        JdbcConnection topmostProxy = getProxy();
-        if (topmostProxy != this.thisAsConnection) {
-            conn.setProxy(this.thisAsConnection); // First call sets this connection as underlying connection parent proxy (its creator).
+    protected ConnectionImpl createConnectionForHost(HostInfo hostInfo) throws SQLException {
+        synchronized (this.lockObject) {
+            ConnectionImpl conn = (ConnectionImpl) ConnectionImpl.getInstance(hostInfo);
```

Comment:
Can we final everything? In IntelliJ you can just go to `Code -> Inspect Code` on the current file, it will a warning for `Declaration can have final modifier`, you can just click `Make final` to fix everything.

Reply:
I'd prefer to do it in a separate PR